### PR TITLE
Add static type checking with MyPy

### PR DESCRIPTION
**Pull request by cthoyt**
# Summary
 
This PR adds optional static type checking using [`mypy`](http://mypy-lang.org/), which uses the type hints to check for errors that don't show up from unit tests.
 
- [x] Code passes all tests
- [x] Unit tests provided for these changes
- [x] Documentation and docstrings added for these changes

## Changes 

1. Add `mypy` environment to `tox.ini`
2. Add call to mypy via tox in the GHA configuration for CI
3. Make a few improvements to code based on suggestions from type checker
4. Add additional tests

### CODE REVIEW

File: rexmex/metricset.py [hunk 1]
```diff
@@ -45,6 +45,8 @@ def filter_metrics(self, filter: List[str] = None):
         Returns:
             self: The metric set after the metrics were filtered out.
         """
+        if not filter:
```
Comment (cthoyt) [inline rexmex/metricset.py:48]: Is this desired functionality? Should it even be possible to call this function with no argument?

Comment (benedekrozemberczki) [reply rexmex/metricset.py:48]: No it is not. I will correct this.

**Comment by benedekrozemberczki** [comment]
Now it conflicts with the main branch.

**Comment by benedekrozemberczki** [comment]
Thank you!

**Comment by cthoyt** [comment]
You're very welcome! Just an explanation for my interest - I might want to integrate RexMex directly in PyKEEN (or at least provide an example of using it an alternative [`pykeen.evaluation.Evaluator`](https://pykeen.readthedocs.io/en/stable/api/pykeen.evaluation.Evaluator.html?highlight=Evaluator#pykeen.evaluation.Evaluator). The only issue I see at the moment is the top-level interface is through pandas dataframes instead of just arrays, which would require writing a compatibility layer to access the more core functionality

**Comment by benedekrozemberczki** [comment]
What would help you? What would be a good API on our side?

**Comment by cthoyt** [comment]
I'm going to start a new issue to continue discussion on this

**Pull request merged**